### PR TITLE
RSDK-4707 Berry IMU module

### PR DESCRIPTION
**Pull request by oliviamiller**
still need to add readme, validation, and rustutils for other OS, these will be in future PRs.  Sorry for the long PR.

### CODE REVIEW

File: libviam_rust_utils.so
Comment (randhid) [inline libviam_rust_utils.so:1]: is this binary necessary?

File: src/berryimu.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import threading
+import asyncio
+from .orientation_vector import euler_angles_to_orientation_vector
+
+
+LOGGER = getLogger(__name__)
+dt = 0.08
+
+class Berryimu(MovementSensor):
+    MODEL: ClassVar[Model] = Model(ModelFamily("viam", "movement_sensor"), "berryimu")
```
Comment (randhid) [inline src/berryimu.py]: your model namespace needs to match the namespace that you're going to upload it in.

File: src/berryimu.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    euler_angles: Vector3
+    calibrate: bool
+    
+
+    @classmethod
+    def new(cls, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]) -> Self:
+        berryimu = cls(config.name)
+        berryimu.reconfigure(config, dependencies)
+        b = threading.Thread(name='readings', target=berryimu.read)
+        b.start()
```
Comment (randhid) [inline src/berryimu.py]: we might want to check the success of your start method, have you seen any instances where it does not start correctly? (i2c pins not connected for example)

File: src/berryimu.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        values = utils.parse_output(raw_data)
+        if self.calibrate:
+            # hard iron distortion correction
+            values.x -= (self.hard_iron_x_min + self.hard_iron_x_max) /2
+            values.y -= (self.hard_iron_y_min + self.hard_iron_y_max) /2
+
+            # #soft iron distortion correction 
+            scale_y = self.soft_iron_y_max - self.soft_iron_y_min
+            scale_x = self.soft_iron_x_max - self.soft_iron_x_min 
+            ratio = scale_y / scale_x
```
Comment (randhid) [inline src/berryimu.py:276]: will scale_x ever be 0? if so we should guard against that

File: src/berryimu.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        values.x = values.x * math.cos(self.euler_angles['pitch']) + values.z * math.sin(self.euler_angles['pitch'])
+
+        values.y = values.x * math.sin(self.euler_angles['roll'])  * math.sin(self.euler_angles['pitch']) + values.y * math.cos(self.euler_angles['roll']) 
+        - values.z * math.sin(self.euler_angles['roll']) * math.cos(self.euler_angles['pitch'])
+
+        heading = 180 * math.atan2(values.y, values.x) / math.pi
+        if heading < 0:
+              heading += 360
+
+        return heading
```
Comment (randhid) [inline src/berryimu.py:299]: confirming this goes 0 north, 90 east, 180 south, 270 west?

File: src/berryimu.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        
+    def get_acceleration(self):
+       raw_data = self.i2cbus.read_i2c_block_data(constants.AG_ADDRESS, constants.OUTX_L_XL, 6)
+       self.accelerometer_raw = utils.parse_output(raw_data)
+      
+       #convert the raw readings to acceleration in m/s^2
+       self.acceleration.x =  self.accelerometer_raw.x * constants.A_GAIN / 1000 * 9.8065
```
Comment (randhid) [inline src/berryimu.py]: is the gravity value based on a s specific coordinate in the world?

File: src/berryimu.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+       # convert raw readings to velocity in degrees/s
+       self.velocity.x = values.x *  constants.G_GAIN
+       self.velocity.y = values.y *  constants.G_GAIN
+       self.velocity.z = values.z *  constants.G_GAIN
+
+       return self.velocity
+
+    def get_gyro_angles(self):
+         # multiply the velocity by the time between readings to get the angle moved. 
+        self.gyro_angles.x = math.radians(self.velocity.x * dt)
```
Comment (randhid) [inline src/berryimu.py]: where is dt defined?

File: src/calibration.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
+# This script provides the values needed to calibrate the IMU's magnetometer
+# Provides parametetrs to correct for hard and soft iron distortions 
+
+import time
+from smbus import SMBus
+import constants
+import utils 
+
+hard_iron_x_min = 32767
```
Comment (randhid) [inline src/calibration.py:10]: a comment on where these values come from.

File: src/calibration.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+         hard_iron_x_min = values.x
+        if values.y > hard_iron_y_max:
+         hard_iron_y_max = values.y
+        if values.y < hard_iron_y_min:
+         hard_iron_y_min = values.y
+        time.sleep(0.025)
+
+# with hard iron distortion applied find the new min and max
+print("finding calibration values for soft iron distortion....keep rotating")
+t_end = time.time() + 30
```
Comment (randhid) [inline src/calibration.py:56]: tell the user they have x seconds to complete calibration

File: src/constants.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+#constants for the LIS3MDL accelerometer and gyroscope
```
Comment (randhid) [inline src/constants.py]: this file could use more explanation in the comments

File: src/main.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+from .berryimu import Berryimu
+
+async def main():
+    module = Module.from_args()
+    module.add_model_from_registry(MovementSensor.SUBTYPE, Berryimu.MODEL)
+    await module.start()
+
+if __name__ == "__main__":
+        asyncio.run(main())
+
```
Comment (randhid) [inline src/main.py]:
```suggestion
        
        
```

File: src/orientation_vector.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import ctypes
+
+from ctypes import Structure, POINTER, c_double
+from viam.proto.common import Orientation
+from viam.logging import getLogger
+
+lib = ctypes.cdll.LoadLibrary("./libviam_rust_utils.so")
```
Comment (randhid) [inline src/orientation_vector.py:7]: ah. now i see.

File: src/utils.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
+from . import constants 
+from viam.proto.common import Vector3
+
+def init_accelerometer(i2cbus):
```
Comment (randhid) [inline src/utils.py:39]: more comments on what these functions do - can you use pylint to see what is suggested for formatting and docstrings?

**Review by randhid — CHANGES_REQUESTED** [review]

File: src/berryimu.py (hunk 3 above)
Comment (oliviamiller) [reply src/berryimu.py:276]: only if the max and min are equal, which would only happen if something went very wrong during the calibration.

File: src/berryimu.py (hunk 5 above)
Comment (oliviamiller) [reply src/berryimu.py]: I used the same value that we are using in the witIMU driver, so I think this is for NYC. can add a comment

File: src/berryimu.py (hunk 6 above)
Comment (oliviamiller) [reply src/berryimu.py]: top of this file, can also move to constants if thats better.

File: src/berryimu.py (hunk 3 above)
Comment (randhid) [reply src/berryimu.py:276]: then let's just make sure to validate against that case in the validator function and tell the user to recalibrate in that instance 👯

File: src/berryimu.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+    def calculate_compass_heading(self):
+        raw_data = self.i2cbus.read_i2c_block_data(constants.MAG_ADDRESS,  constants.OUT_X_L, 6)
+        values = utils.parse_output(raw_data)
+        if self.calibrate:
+            # hard iron distortion correction
+            values.x -= (self.hard_iron_x_min + self.hard_iron_x_max) /2
+            values.y -= (self.hard_iron_y_min + self.hard_iron_y_max) /2
+
+            # #soft iron distortion correction
```
Comment (martha-johnston) [inline src/berryimu.py]: extremely small nit: remove the second #

File: src/berryimu.py [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+        self.accelerometer_angles.y= math.asin(acc_y_norm/math.cos(self.accelerometer_angles.x))
+
+        return self.accelerometer_angles
+    
+    def calculate_orientation(self):
+        self.accelerometer_angles = self.get_acc_angles()
+        self.gyro_angles = self.get_gyro_angles()
+
+
+        ## complementary fiter is used to combine the readings from the gyro and accelerometer.
```
Comment (martha-johnston) [inline src/berryimu.py]: another double #

**Review by martha-johnston — APPROVED** [review]
agree with rand's comments about where to add more descriptive comments, but other than that lgtm!

File: src/berryimu.py (hunk 1 above)
Comment (oliviamiller) [reply src/berryimu.py]: Oh I see so make it viam-labs? How do I know if I have access to a namespace?

File: src/berryimu.py (hunk 2 above)
Comment (oliviamiller) [reply src/berryimu.py]: will do some more testing

Comment (oliviamiller) [reply src/berryimu.py]: in the case that the i2c pins are not connected correctly reconfigure will fail and this thread will never start, eventually you get `configuration error: reached max of 5 configuration attempts for rdk:component:movement_sensor/imu `. What should the expected behavior be in this case?

Comment (randhid) [reply src/berryimu.py]: That's expected behavior - add the readme that the modules expect the pins to be connected at startup.

File: run.sh [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+# -qq suppresses extraneous output from pip
+echo "Virtualenv found/created. Installing/upgrading Python packages..."
+if ! $PYTHON -m pip install -r requirements.txt -Uqq; then
+    exit 1
+fi
+
+# Be sure to use `exec` so that termination signals reach the python process,
+# or handle forwarding termination signals manually
+echo "Starting module..."
+exec $PYTHON -m src.main $@
```
Comment (randhid) [inline run.sh]:
```suggestion
exec $PYTHON -m src.main $@

```

File: src/berryimu.py (hunk 1 above)
Comment (randhid) [reply src/berryimu.py]: you can try uploading anythign to a namespace through cli, but also - you have access to any orgs that you are an owner of in app.viam.com, their public namespaces will be available to you onc you log into the cli. 

I'm also okay if you want to upload it under your own namespace.

**Review by randhid — APPROVED** [review]
LGTM! Upload it to viam-labs or to your own namespace when ready. Did you manage to fix the import errors on the python packages?

Also - make sure al dependencies we want are in the requirements file! Try testing on a pi but adding the module after this is merged in

**Comment by oliviamiller** [comment]
> LGTM! Upload it to viam-labs or to your own namespace when ready. Did you manage to fix the import errors on the python packages?
> 
> Also - make sure al dependencies we want are in the requirements file! Try testing on a pi but adding the module after this is merged in

will upload to viam-labs namespace since this is in viam-labs github org! yes I just fixed by combining constants and utils into one file. I think the only two dependencies are in requirements.txt already but will test again to be sure!

**Pull request merged**